### PR TITLE
first commit for the addition of the TabDDPM plugin

### PR DESCRIPTION
**Pull request by TZTsai**
## Description
I added `TabDDPMPlugin`, named "ddpm", in the generic plugin category, implementing the Tabular DDPM model proposed in [this paper](https://arxiv.org/pdf/2209.15421.pdf). A `torch.nn.Module`, `TabDDPM`, is implemented in the folder "src\synthcity\plugins\core\models\tabular_ddpm". The test cases for the "ddpm" plugin are written in "tests\plugins\generic\test_ddpm.py" accordingly. I also added a `discrete_columns` function in "src\synthcity\utils\dataframe.py" to determine the discrete features in a dataframe by its number of unique values. This is mainly to help the implementation of `TabDDPM`, but I also used this function to simplify the code in "src\synthcity\plugins\core\dataloader.py".

## How has this been tested?
I ran the `test_ddpm.py` file and all 24 test cases were passed.
图片

## Checklist
- [X] I have followed the [Contribution Guidelines](https://github.com/vanderschaarlab/.github/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/vanderschaarlab/.github/blob/main/CODE_OF_CONDUCT.md)
- [X] I have commented my code following the [van der Schaar Lab Styleguide](https://github.com/vanderschaarlab/.github/blob/main/STYLEGUIDE.md)
- [X] I have labeled this PR with the relevant [Type labels](https://github.com/vanderschaarlab/.github/labels?q=Type%3A)
- [X] My changes are covered by tests

### CODE REVIEW

**Review by bcebere — CHANGES_REQUESTED** [review]
Hello @TZTsai 

Thank you for your work.

Before starting the review, please fix the code to pass the static checkers.

You can do that locally by running 
```
pre-commit install
pre-commit run --all
```
and reviewing the errors.

Thank you!

**Comment by ZhaozhiQIAN** [comment]
Hi @robsdavis , could you please work with Bogdan to review & moderate this pull request? Hi @bcebere , please help Rob to get familiar with the pull request reviewing process (as an important step of handover). 

Many thanks!

File: src/synthcity/plugins/core/models/tabular_ddpm/__init__.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/__init__.py:1]: You shouldn't disable the existing checks. All the other modules respect these checks. Please fix the issues instead.

**Review by bcebere — CHANGES_REQUESTED** [review]

File: src/synthcity/plugins/core/models/tabular_ddpm/utils.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/utils.py:26]: Having asserts in public code it is not safe. Replace this with exceptions instead

**Review by bcebere — CHANGES_REQUESTED** [review]

File: src/synthcity/plugins/core/models/tabular_ddpm/__init__.py
Comment (TZTsai) [reply src/synthcity/plugins/core/models/tabular_ddpm/__init__.py]: I have removed all these comments except flake8 F405, which prohibits star import. I have a utils.py and another internal script imports a lot of variables and functions from it. I wonder if it is also mandatory that I need to replace the star import with a long list of names.

File: src/synthcity/plugins/core/models/tabular_ddpm/utils.py
Comment (TZTsai) [reply src/synthcity/plugins/core/models/tabular_ddpm/utils.py]: I have replace all the asserts.

File: src/synthcity/plugins/core/models/tabular_ddpm/__init__.py
Comment (bcebere) [reply src/synthcity/plugins/core/models/tabular_ddpm/__init__.py]: Hello @TZTsai 

Thank you so much for working on fixing the errors!
It would be recommended to import by name instead of star import. Otherwise, you could import by mistake something that hides a name in the current module, without knowing that.

Thanks again for the contribution!

Comment (TZTsai) [reply src/synthcity/plugins/core/models/tabular_ddpm/__init__.py]: Hello @bcebere 

You are welcome! I have removed star import now as well.

File: src/test.py
Comment (bcebere) [inline src/test.py:1]: Is this script working? Because it looks like the `ddpm` cannot be loaded in tests.

Also, please remove this file from the git, keep it for local tests only.

File: src/synthcity/plugins/generic/plugin_ddpm.py
Comment (bcebere) [inline src/synthcity/plugins/generic/plugin_ddpm.py:4]: Please re-enable the checks here as well. Both mypy and flake8

The mypy output gives you an indication why the tests are failing: the `_fit` method has a different signature than the Plugin interface. you need to stick to `def _fit(self, X: DataLoader, *args: Any, **kwargs: Any)`. This is why the tests are failing.

**Review by bcebere — CHANGES_REQUESTED** [review]

File: src/synthcity/plugins/generic/plugin_ddpm.py
Comment (TZTsai) [reply src/synthcity/plugins/generic/plugin_ddpm.py]: Sorry I didn't notice that. I have changed the signatures.

Comment (bcebere) [reply src/synthcity/plugins/generic/plugin_ddpm.py]: I had a look why this still fails.

It seems like for Python3.7 or 3.8 you cannot use subscripts `Iterator`. See the diff below.

After that, the tests should run fine, and I will do an in-depth review 

```diff
diff --git a/src/synthcity/plugins/core/models/tabular_ddpm/__init__.py b/src/synthcity/plugins/core/models/tabular_ddpm/__init__.py
index 0618bec..3591000 100644
--- a/src/synthcity/plugins/core/models/tabular_ddpm/__init__.py
+++ b/src/synthcity/plugins/core/models/tabular_ddpm/__init__.py
@@ -55,8 +55,8 @@ class TabDDPM(nn.Module):
 
     def _update_ema(
         self,
-        target_params: Iterator[nn.Parameter],
-        source_params: Iterator[nn.Parameter],
+        target_params: Iterator,
+        source_params: Iterator,
         rate: float = 0.999,
     ) -> None:
         """
         ```

File: src/test.py
Comment (TZTsai) [reply src/test.py]: I met this problem of failing to load the new plugin when I tested it for the first time as well. What I did was to use pip to reinstall the project again `pip install -U .`. Is this problem during the checking due to the same reason?

Comment (bcebere) [reply src/test.py]: See my comment below. The problem is related to Py3.7/3.8

File: src/synthcity/plugins/core/models/tabular_ddpm/utils.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/utils.py:69]: Please use `ValueError` with a meaningful error message.

File: src/synthcity/plugins/core/models/tabular_ddpm/utils.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/utils.py:87]: Please use `ValueError` with a meaningful error message.

File: src/synthcity/plugins/core/models/tabular_ddpm/utils.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/utils.py:130]: `extract` is a too generic name. Please write something more specific.

File: src/synthcity/plugins/core/models/tabular_ddpm/utils.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/utils.py:201]: Please use `ValueError` with a meaningful error message.

File: src/synthcity/plugins/core/models/tabular_ddpm/utils.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/utils.py:181]: Shouldn't this inherit ` torch.utils.data.DataLoader`?

File: src/synthcity/plugins/core/models/tabular_encoder.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_encoder.py:107]: `BinEncoder` works on Series. Are you using it in DDPM?

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:98]: Instead of the `verbose` arg, please use the `synthcity.logger` module, with `log.error`/`log.debug` etc. You can find examples in the code

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:112]: Please use `synthcity.logger`

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:204]: Please use ValueError with a meaningful error message.

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:209]: Please use ValueError with a meaningful error message.

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:211]: Please use ValueError with a meaningful error message.

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:261]: Please use ValueError with a meaningful error message.

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:286]: Please use ValueError with a meaningful error message.

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:303]: Please use ValueError with a meaningful error message.

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:369]: Please use ValueError with a meaningful error message.

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:504]: Please use ValueError with a meaningful error message.

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:527]: If this commented code is not useful, please delete the comment

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:696]: If the commented code is not useful, please delete it

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:767]: Delete the commented code

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:756]: Delete the commented code if not useful

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:832]: Delete the commented code if not useful

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:878]: Delete the commented code if not useful

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:973]: Delete the commented code if not useful

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:1020]: Add a meaningful error message

File: src/synthcity/plugins/core/models/tabular_ddpm/modules.py
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabular_ddpm/modules.py:118]: There is another MLP in the codebase, at synthcity.plugins.core.models.mlp

Consider reusing that class instead, and drop this one

File: src/synthcity/plugins/generic/plugin_ddpm.py
Comment (bcebere) [inline src/synthcity/plugins/generic/plugin_ddpm.py:34]: Please document the args

File: src/synthcity/plugins/generic/plugin_ddpm.py
Comment (bcebere) [inline src/synthcity/plugins/generic/plugin_ddpm.py:168]: `dataframe` includes the target column as well

File: src/synthcity/plugins/generic/plugin_ddpm.py
Comment (bcebere) [inline src/synthcity/plugins/generic/plugin_ddpm.py:174]: Why are you not using the tabular encoding?

File: tests/plugins/generic/test_ddpm.py
Comment (bcebere) [inline tests/plugins/generic/test_ddpm.py:18]: Drop the commented code, if not useful

**Review by bcebere — CHANGES_REQUESTED** [review]
@TZTsai Amazing work so far, congrats for the effort!

I did a more in-depth review, see my comments below.

At the same time, please add a notebook tutorial in `https://github.com/vanderschaarlab/synthcity/tree/main/tutorials/plugins/generic`

Thanks!

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (TZTsai) [reply src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:407]: I think this would have a problem that if we modify `model_kwargs` inplace, this will change the default value of `model_kwargs`.

File: src/synthcity/plugins/generic/plugin_ddpm.py
Comment (TZTsai) [reply src/synthcity/plugins/generic/plugin_ddpm.py:168]: Yes, but I was considering if we should drop the target column if it is already provided as `cond`. Sorry I should not say "include".

File: src/synthcity/plugins/generic/plugin_ddpm.py
Comment (TZTsai) [reply src/synthcity/plugins/generic/plugin_ddpm.py:174]: Because the Gaussian multinomial diffusion module needs to know the number of classes for each discrete feature before it applies `torch.nn.functional.one_hot` on these features, and it also preprocesses the continuous features differently.

Comment (bcebere) [reply src/synthcity/plugins/generic/plugin_ddpm.py:174]: OK, then remove this comment, and leave a comment in the code explaining why the TabularEncoding is not needed.

File: src/synthcity/plugins/core/models/tabular_encoder.py
Comment (TZTsai) [reply src/synthcity/plugins/core/models/tabular_encoder.py:107]: But in Line 119 and 254 there are for name in raw_data.columns. Are they not supposed to be Dataframes?

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (bcebere) [reply src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:407]: But for every call that uses `model_kwargs`, you do

``` 
if model_kwargs is None:
    model_kwargs = {}
```
            anyway

File: src/synthcity/plugins/core/models/tabular_encoder.py
Comment (bcebere) [reply src/synthcity/plugins/core/models/tabular_encoder.py:107]: Great catch! Thanks for fixing this.

File: src/synthcity/plugins/core/models/tabular_ddpm/utils.py
Comment (TZTsai) [reply src/synthcity/plugins/core/models/tabular_ddpm/utils.py:181]: This has a different signature from the torch DataLoader. More specifically, its input is not a Dataset but the tensors instead. It is said by the original author to be much faster than the torch implementation. We can remove it and use the standard DataLoader instead.

File: src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py
Comment (TZTsai) [reply src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py:407]: Yes, but the difference is that if the default `model_kwargs` is `{}`, then when I add an item in this dict, e.g `model_kwargs[k] = v`, the next time I can this method, the default value of `model_kwargs` is no longer `{}` but `{k: v}`. Actually it will create no problem as long as I do not mess around with the default value. This is just a matter of habit to ensure that the default value never changes.

File: src/synthcity/plugins/core/models/tabular_ddpm/utils.py
Comment (TZTsai) [reply src/synthcity/plugins/core/models/tabular_ddpm/utils.py:174]: I have replaced it with the standard combination of TensorDataset and DataLoader.

**Comment by TZTsai** [comment]
@bcebere @ZhaozhiQIAN 

I have added a file "synthcity/utils/callbacks.py" and created an `EarlyStopping` callback, which a plugin can use it in a callback list in the arguments of initialization. To use `EarlyStopping`, a `torch.nn.Module` must have an `X_val` dataset and call `callback.on_epoch_end(self)` to determine whether to terminate the training loop. If it considers the training should be stopped early, a StopIteration exception will be raised and the model should catch it. Finally, the model can call `callback.on_fit_end(self)` to load the state dict of the best model.

**Review by bcebere — COMMENTED** [review]
@TZTsai Thank you again for the incredible PR.

One last request: please add an tutorial using this model in tutorials/plugins/generic.
Also, Update the table in the README.md with this new method.

@robsdavis if this looks good to you, after the tutorials are added, feel free to merge this PR into main.

**Review by robsdavis — APPROVED** [review]
This looks like some really great work! I will merge once you have a tutorial ready.

**Comment by pravsels** [comment]
Hey all. I wanted to try this method out, so I was preparing a tutorial to get this PR merged but found a couple of issues in 'modules.py' in tabular_ddpm folder. Was able to fix it and train the model locally. 

There are still issues with the generate flow which I'm hoping to get to later. 

In the meantime, I'm not too sure on how I could push my changes. Although my fork is up to date, it doesn't have the changes from TZTsai's fork and I can't create another fork from their fork. 

Any ideas ? @robsdavis @bcebere

**Comment by TZTsai** [comment]
> Hey all. I wanted to try this method out, so I was preparing a tutorial to get this PR merged but encountered a couple of issues (one with tensors not being on the same device and another with resize/reshape) in 'modules.py' in tabular_ddpm folder. Was able to fix it and train the model locally.
> 
> There are still issues with the generate flow (about indices being on the same device as the tensor being indexed) which I'm hoping to get to later.
> 
> In the meantime, I'm not too sure on how I could push my changes. Although my fork is up to date, it doesn't have the changes from TZTsai's fork and I can't create another fork from their fork.
> 
> Any ideas ? @robsdavis @bcebere

Sorry I have been obsessed with GPT4 recently, since I have got its API access and I wrote a Telegram bot to chat with it. I can write the tutorial this weekend, but if you are really interested you are welcome to push the changes.

**Comment by bcebere** [comment]
> > Hey all. I wanted to try this method out, so I was preparing a tutorial to get this PR merged but encountered a couple of issues (one with tensors not being on the same device and another with resize/reshape) in 'modules.py' in tabular_ddpm folder. Was able to fix it and train the model locally.
> > There are still issues with the generate flow (about indices being on the same device as the tensor being indexed) which I'm hoping to get to later.
> > In the meantime, I'm not too sure on how I could push my changes. Although my fork is up to date, it doesn't have the changes from TZTsai's fork and I can't create another fork from their fork.
> > Any ideas ? @robsdavis @bcebere
> 
> Sorry I have been obsessed with GPT4 recently, since I have got its API access and I wrote a Telegram bot to chat with it. I can write the tutorial this weekend, but if you are really interested you are welcome to push the changes.

Hello both,

Thank you again  for your contributions. @robsdavis will merge this PR tomorrow, and then you are welcome to add more tutorials in a separate PR. Thank you!

**Review by bcebere — APPROVED** [review]

**Pull request merged**
